### ISSUE TITLE
System.in, System.out, System.err fields are incorrectly treated as immutable in Java

### ISSUE THREAD

**Issue by wandernauta** [utwente-fmt/vercors#1242]
The following Java class verifies:

```java
import java.io.InputStream;
import java.lang.System;

class Example {
    public static void example(InputStream foo) {
        if (System.in != null && foo != null) {
            InputStream a = System.in;
            System.setIn(foo);
            InputStream b = System.in;

            assert(a == b);
        }
    }
}
```

However, the assertion does not hold in real programs for [legacy reasons, per JLS](https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.5.4); these three specific fields have [very special](https://github.com/openjdk/jdk/blob/master/src/java.base/share/native/libjava/System.c#L258C37-L259C54) "a bit `static final` but not _really_ `static final`" behavior that the compiler is aware of.

**Comment by bobismijnnaam** [comment]
One way to fix this would be to encode `in`, `out` and `err` as methods which have only `\result != null` as postcondition. This reflects real world behaviour nicely, meaning not being allowed to use them in contracts as they can change at any time, and assuming they change if you refer to them twice. It does require some extra fiddling in the frontend to get vercors to recognize these fields as special, but that is to be expected I guess. Too bad it will make the langspecifictocol pass more complicated...

**Comment by wandernauta** [comment]
That makes sense. Though, I was looking at that `\result != null` postcondition as well - the real System class doesn't seem to make that promise, this succeeds (does not verify):

```java
class Example {
    public static void main(String[] args) {
        System.setOut(null);
        assert(System.out == null);
    }
}
```

Is it enough for `setIn`, `setOut`, `setErr` to `@require` this to make sure that nobody ever calls it with `null`?

**Comment by bobismijnnaam** [comment]
Yes, if these setters are really the only way to change the special fields, making the contracts of the setters require non-null arguments should be sufficient.
